lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+    components: Vec<T>,
+}
+
+// Represents a single XML component.
+#[derive(Clone, Debug, Deserialize)]
+struct XmlComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Components<XmlComponent>>,
+}
+
+impl Component for XmlComponent {
+    fn component_type(&self) -> &str {
+        &self.component_type
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn version(&self) -> &str {
+        &self.version
+    }
+
+    fn scope(&self) -> Option<&str> {
+        self.scope.as_deref()
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_ref().map(|comps| comps.components.as_slice())
+    }
+}
+
+// Represents a single JSON component.
+#[derive(Clone, Debug, Deserialize)]
+struct JsonComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Vec<JsonComponent>>,
+}
+
+impl Component for JsonComponent {
+    fn component_type(&self) -> &str {
+        &self.component_type
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn version(&self) -> &str {
+        &self.version
+    }
+
+    fn scope(&self) -> Option<&str> {
+        self.scope.as_deref()
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_deref()
+    }
+}
+
+// Filter components based on their type and scope.
+fn filter_components<T: Component>(components: &[T]) -> impl Iterator<Item = &'_ T> {
+    components
+        .iter()
+        .filter(|&comp| {
+            let type_check = comp.component_type() == "application"
+                || comp.component_type() == "framework"
+                || comp.component_type() == "library";
+
+            // Check if the scope is "required" or not specified (required)
+            let scope_check = match comp.scope() {
+                Some(scope) => scope == "required",
+                None => true,
+            };
+
+            type_check && scope_check
+        })
+        .flat_map(|comp| {
+            let nested_iter = match comp.components() {
+                Some(nested) => filter_components(nested).collect::<Vec<_>>(),
+                None => Vec::new(),
+            };
+            std::iter::once(comp).chain(nested_iter.into_iter())
+        })
+}
+
+// Convert a component's Package URL (PURL) into a Package object.
+fn from_purl<T: Component>(component: &T) -> anyhow::Result<Package> {
+    let purl_str = component
+        .purl()
+        .ok_or_else(|| anyhow!("Missing purl for {}:{}", component.name(), component.version()))?;
```

Comment:
Is this required by specification? Generally if it is, why wouldn't we enforce that in the deserialization?

Providing better error messages is nice, but that's only useful if people are expected to run into them.

And if it's actually optional, should we maybe not error out?

Reply:
It's optional by specification however for us it wouldn't be optional. A similar conversation occurred with SPDX as I initially just ignored packages missing a PURL. The outcome was that we could not determine the ecosystem otherwise and could lead to providing consumers incorrect results (we do ignore ecosystems we don't support), so we fail the parsing and give the user an error.

Reply:
Hm that certainly doesn't seem like a great solution, but I'm not familiar with sboms and there probably just isn't any good resolution.

Reply:
I think we can pass on parsing external refs for now for the first release. There needs to be additional work for the optional dependencies. Can wrap that work into next iteration.